# Summer School on Polyploid evolutionary genomics, 2023
10-20 July
[University of Fribourg](https://events.unifr.ch/summerschool_polyploidy/en/)

![Summer School](https://www.ville-fribourg.ch/sites/default/files/2018-06/vue_planche_sup.jpg)

## Introduction:
Welcome to the Summer School on Polyploid evolutionary genomics, 2023 held at the University of Fribourg, Switzerland! This renku server provides you with a comprehensive guide and platform to follow to our specialized summer school practical session focused on various aspects of polyploidy research. Here, you will find details about the schedule, daily contents, and the esteemed instructors leading the exciting topics of the course.

## Starting the session
1. Use the start button on top right corner to start a new session.

     ![start](/images/start.jpg)
[Note] If you want to save your output and working files you first have to register on Renku, fork this repository and start from your own fork. Once you have progressed through the tutorials and want to save your files, just type `renku save` on the terminal.

2. If you are starting the session for the first time it will take ~10 minutes to download the pre-build image files. So if you are seeing this page for a while, don't panic, your session will load soon.

     ![loading](/images/loading.jpg)

3. Once your session has started you will see the following page. You have the options highlighted here to work on these tutorials. In short for command line tools you can choose either ´Terminal´ or ´VScode´, For R tutorials, choose ´Rstudio´ and for graphical desktop choose ´VNC´.

     ![workspace](/images/workspace.jpg)



## Schedule:

#### Day 1: 
**Afternoon**
- **Date:** July 14, 2023
- **Time:** 13:30
- **Instructor:** [Dr. Quinten Bafort](https://www.vandepeerlab.org/people/qubaf), [Dr. Zhen Li](https://www.vandepeerlab.org/people/zheli)
- **Topic:** Assembly of polyploid genomes and subgenome evolution
- **Notebook:** [<span style="color:red">Link</span>](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023/files/blob/notebooks/Day1/Assembly_of_polyploid_genomes.md)
- **Contents:**
  - Participants will learn about the extra complexities of (the assembly of) polyploid genomes that consist of (very) similar or identical subgenomes. 
  - Participants will learn about how to look for evidence of subgenome dominance, and ongoing rediploidization.
  
#### Day 2: 
**Morning**
- **Date:** July 15, 2023
- **Time:** 8.30
- **Instructor:** [Dr. Zhen Li](https://www.vandepeerlab.org/people/zheli), [Dr. Quinten Bafort](https://www.vandepeerlab.org/people/qubaf)
- **Topic:** Ks age distribution and genomic dot plots
- **Notebook:** [<span style="color:red">Link</span>](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023/files/blob/notebooks/Day2/Ks_age_distribution_and_genomic_dot_plots.md)
- **Contents:**
  - The construction of Ks age distributions and genomic dot plots are two of the most widely used and important ways to find evidence for ancient polyploidy. 
  - Where the construction of Ks distributions is not based on structural information and can be even based on extensive transcriptome information, the construction of genomic dot plots requires structural information and a well-assembled genome.

**Afternoon**
- **Date:** July 15, 2023
- **Time:** 13.30
- **Instructor:** [Dr. Paul Simion](https://ecobio.univ-rennes.fr/interlocuteurs/paul-simion), [Dr. Zhen Li](https://www.vandepeerlab.org/people/zheli), [Dr. Quinten Bafort](https://www.vandepeerlab.org/people/qubaf)
- **Topic:** Gene trees – species tree reconciliation
- **Notebook:** [<span style="color:red">Link</span>](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023/files/blob/notebooks/Day2/Reconciliation_Course_Practicals.md)
- **Contents:**
  - Gene tree reconciliation enables to look for evidence for (very) ancient polyploidy, where this is often not possible with Ks age distributions, because of saturation of synonymous mutations for large evolutionary distances. 
  - Participants will learn to work with software to construct Ks age distributions, to make genomic
dot plots, and to perform gene tree – species tree reconciliation.
  
#### Day 3: 
**Morning**
- **Date:** July 17, 2023
- **Time:** 8.30
- **Instructor:** [Prof. Kentaro Shimizu](https://www.ieu.uzh.ch/en/staff/member/shimizu_kentaro.html), [Dr Masaomi Hatakeyama](https://www.ieu.uzh.ch/en/staff/member/hatakeyama_masaomi.html), [Dr. Moeko Okada](https://moekookada.github.io)
- **Topic:** RNA-seq analysis using subgenome classification methods
- **Notebook:** [<span style="color:red">Link</span>](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023/files/blob/notebooks/Day3/Homeologous_expression_analysis.md)
- **Contents:**
  - For high quality RNA-seq analysis of allopolyploid species, bioinformatic workflows called subgenome classification methods were developed. 
  - We will introduce bioinformatic workflow for subgenome classification methods (HomeoRoq, EAGLE-RC) combined with RNA-seq mapping and quantification tools.
  - We also focus on significant changes in homeolog expression ratio.
  
**Afternoon**
- **Date:** July 17, 2023
- **Time:** 13.30
- **Instructor:** [Prof. Kentaro Shimizu](https://www.ieu.uzh.ch/en/staff/member/shimizu_kentaro.html), [Dr Masaomi Hatakeyama](https://www.ieu.uzh.ch/en/staff/member/hatakeyama_masaomi.html), [Dr. Moeko Okada](https://moekookada.github.io)
- **Topic:** Robustness of polyploid species
- **Contents:**
  - Recent studies using several allopolyploid species showed that allopolyploid species inherited and combined the gene expression patterns of progenitor species. 
  - We suggest that they contributed to the environmental robustness of allopolyploid species.
  
### Day 4: 
**Morning**
- **Date:** July 18, 2023
- **Time:** 8.30
- **Instructor:** [Dr. Armel Salmon](https://ecobio.univ-rennes.fr/en/interlocutors/armel-salmon), [Dr. Rimjhim Roy Choudhury](https://rimjhimroy.github.io), [Prof. Christian Parisod](https://www.unifr.ch/bio/en/department/staff/prof/people/334843/93f3c)
- **Topic:** Genotyping polyploids
- **Notebook:** [<span style="color:red">Link</span>](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023/files/blob/notebooks/Day4/SNP_calling_polyploids.md)
- **Contents:**
  - Genotyping polyploids is challenging. Participants will learn about avoidable pitfalls when calling for single nucleotide polymorphisms as well as structural variants among polyploid samples and about strategies to accurately assess (fixed or partial) heterozygotes. 
  
**Afternoon**
- **Date:** July 18, 2023
- **Time:** 13.30
- **Instructor:** [Dr. Sandra Grünig](https://www.unifr.ch/directory/fr/people/342164/fd61e), [Prof. Christian Parisod](https://www.unifr.ch/bio/en/department/staff/prof/people/334843/93f3c)
- **Topic:** Autopolyploid populations
- **Notebook:** [<span style="color:red">Link</span>](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023/files/blob/notebooks/Day4/Autopolyploid_populations.md)
- **Contents:**
  - Participants will learn how to assess tetrasomic inheritance and to compare the linkage and structure of genomic variation within and among populations of diploid-autotetraploid systems. 
  
### Day 5:
**Morning** 
- **Date:** July 19, 2023
- **Time:** 8.30
- **Instructor:** [Prof. Martin Lascoux](https://www.katalog.uu.se/profile?id=N96-1825), [Dr Mathieu Tiret](https://www6.val-de-loire.inrae.fr/biofora/Personnel/Ils-ont-travaille-a-BioForA/TIRET-Mathieu)
- **Topic:** Allopolyploid populations
- **Notebook:** [<span style="color:red">Link</span>](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023/files/blob/notebooks/Day5/Polyploidy_Cbp.Rmd)
- **Contents:**
  - Participants will learn how to deal with genomic variation in allopolyploid systems that include wholegenome duplication and hybridization, with the latter that seems to matter more than the former, at least for expression. 
  
**Afternoon**
- **Date:** July 19, 2023
- **Time:** 13.30
- **Instructor:** [Dr Mathieu Tiret](https://www6.val-de-loire.inrae.fr/biofora/Personnel/Ils-ont-travaille-a-BioForA/TIRET-Mathieu), [Prof. Martin Lascoux](https://www.katalog.uu.se/profile?id=N96-1825)
- **Topic:** Allopolyploid populations
- **Contents:**
  - Participants will learn that parental legacy plays an important role and seems to have a rather long-lasting effect. 
  - So to understand an allopolyploid we need to understand its parental species. Although allopolyploid subgenomes have been in the same individuals for a long time and therefore have had the same demography they still differ significantly.
  
### Day 6: 
**Morning**
- **Date:** July 19, 2023
- **Time:** 8.30
- **Instructor:** [Prof. Martin Lascoux](https://www.katalog.uu.se/profile?id=N96-1825), [Dr Mathieu Tiret](https://www6.val-de-loire.inrae.fr/biofora/Personnel/Ils-ont-travaille-a-BioForA/TIRET-Mathieu)
- **Topic:** Signatures of selection
- **Contents:**
  - Participants will learn to analyze patterns of selection and highlight selective sweeps in allopolyploids.
  
**Afternoon**
- **Date:** July 19, 2023
- **Time:** 8.30
- **Instructor:** [Prof. Martin Lascoux](https://www.katalog.uu.se/profile?id=N96-1825), [Prof. Christian Parisod](https://www.unifr.ch/bio/en/department/staff/prof/people/334843/93f3c)
- **Topic:** Interpretations and conclusions”
- **Contents:**
  - Participants will revisit hypotheses formulated at the beginning of the summer school in light of the approaches they have been practicing and highlight perspectives for polyploidy research at the population level. 
  
  
  
This is the Renku project for the [Polyploid Summer School 2023](https://events.unifr.ch/summerschool_polyploidy/en/). Renku is a platform developed by the Swiss Data Science Center. It is a joint effort of ETH Zurich and EPFL Lausanne to provide reproducible computational environments for collaborating on data science research for teaching. Renku project is basically a git repository with some bells and whistles. You'll find we have already created some useful things like `data` and `notebooks` directories and a `Dockerfile`.


## Quick Renku instructions to work on a forked repository:

1. Go to [https://renkulab.io/](https://renkulab.io/). Click on the 'Login or Sign Up' button and make yourself an account if you do not have one, either via 'Register', a 'SWITCH edu-ID' or via a GitHub account.
2. Once logged in, go to the Polyploid Summer School "project":
[https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023](https://renkulab.io/projects/rimjhim.choudhury/polyploid-summer-school-2023)
3. Click on the button to fork the project; this should bring up a box where you can modify the name (you don't need to) and you can click on the 'Fork' button, putting you on the landing page of your (new) project.
4. There are a few ways of initiating a session. The one-click Start button uses all the default settings to initiate a session. You can also click on the drop-down menu next to the start and select "stat with options". Here is a detailed guide on how to start an interactive session: https://renku.readthedocs.io/en/latest/tutorials/first_steps/02_start_an_interactive_session.html 
5. Once you click on start, hopefully it shows "Docker image available"; if not, perhaps it says "Docker Image building", which is also ok (it might take a couple of minutes to build and then turn to "Docker image available"). 

Note: you can always go back and forth between the `/rstudio`, `/lab`, and `/vscode` environments by modifying the end of web link, e.g., if you started in /rstudio, but want to move to /lab, you can remove "rstudio", type "lab" in its place and then press enter.

Note: you can upload files from the local computer into your environment via the left tab of the /lab/ environment.


## Working with Renku

Please see [the documentation](https://renku.readthedocs.io/en/latest/) for more details about Renku.





